optuna/trial/_trial.py

```diff
@@ -384,14 +388,23 @@ def objective(trial):
                 Lower endpoint of the range of suggested values. ``low`` is included in the range.
             high:
                 Upper endpoint of the range of suggested values. ``high`` is included in the range.
+            log:
+                A flag to sample the value from the log domain or not.
+                If ``log`` is true, the value is sampled from the range in the log domain.
+                Otherwise, the value is sampled from the range in the linear domain.
             step:
                 A step of spacing between values.
 
         Returns:
             A suggested integer value.
         """
 
-        distribution = distributions.IntUniformDistribution(low=low, high=high, step=step)
+        if log:
+            distribution = IntLogUniformDistribution(
+                low=low, high=high, step=step
+            )  # type: Union[IntUniformDistribution, IntLogUniformDistribution]
+        else:
+            distribution = IntUniformDistribution(low=low, high=high, step=step)
```

Comment:
I think we should also add type hinting here. 
```suggestion
            distribution = IntUniformDistribution(
                low=low, high=high, step=step
            ) # type: Union[IntUniformDistribution, IntLogUniformDistribution]
```

Reply:
It may not be possible.  When applying your suggestion, mypy causes the following error:

```
optuna/trial/_trial.py:407: error: Name 'distribution' already defined on line 403
```

Reply:
Sorry for the confusion. I think the following would be better because it is easy to understand why it has a type of `Union[IntUniformDistribution, IntLogUniformDistribution]`.
```suggestion
        distribution = IntUniformDistribution(
            low=low, high=high, step=step
        ) # type: Union[IntUniformDistribution, IntLogUniformDistribution]
        if log:
            distribution = IntLogUniformDistribution(low=low, high=high, step=step)
```

Reply:
sorry for my misunderstanding and thank you for your clarification. Looks nice! I merge your suggestion.